### PR TITLE
Robuster Build Environment

### PR DESCRIPTION
**Pull request by td-usds**
This PR makes our build environment a bit robuster than it is, and it aids new developers in coming on board and having a functional baseline. It does this in two ways:
1. provide a container for the CI/CD pipeline to perform the build in and which you can use yourself as well to do a build as if you were the CI/CD pipeline (if you so choose) - note that the container is used in the `docker-compose.build.yml` file and pulled from https://github.com/orgs/CDCgov/packages/container/package/prime-router_builder (which is our own Container Registry on GitHub)
2. provide a `cleanslate.sh` script to clean your environment and get back to a known good state or to get your first build going asap as a new developer; this script will also set everything up for a new developer who just did a clone (and installed the prerequisites as per the getting-started.md file)

### ./prime-router
- streamlining of the `getting_started.md` document (pdf of the new content included in this PDF:
[getting started.pdf](https://github.com/CDCgov/prime-reportstream/files/6793391/getting.started.pdf))

- This **adds a prerequisite that docker & docker-compose is installed** on the developer's machine. 

- The changes require that you run your PostgreSQL as a docker container and not on 'bare metal'.

- Adds a docker-compose environment (`docker-compose.build.yml`) which provides a standardized build environment for the product. This environment builds the builder as defined in `Dockerfile.build`.

- Removes the `docker-infrastructure.yml` docker-compose environment (as it is now supplanted by `docker-compose.build.yml`); note that the new build docker-compose does NOT include `adminer` to keep it more 'lightweight'. Anyone who wants to put an admin UI on top of the (exposed) PGSQL can just point at their own `localhost:54342` with the tool of their choice.

- Adds a `./build.sh` script to do the build in the container with the ability to run different build targets; this script can be used on a developer's box to perform a build in a similar environment that the GitHub pipeline performs a build in. You can also bash into the container to do a build as if you are the CI/CD pipeline by doing:
    ```bash
    ./devenv-infrastructure.sh stay # this will drop you into a shell inside the container
    ```

- Adds a `./cleanslate.sh` script (use `./cleanslate.sh --help` for more info) to either clean up your environment in case it got hosed or to get started with everything set up as a new developer

### ./.github/workflows
- Updates the GitHub workflows `build_hub.yml` ("Router") and `release.yml` ("Release to azure") to use the new container-based build mechanism

### What/How to validate
```bash
# Run the cleanslate script in its most aggressive form (see --help and caveat emptor)
cd prime-router
./cleanslate --prune-volumes
# You can now also inspect the cleanslate.sh.log file for more info in case something went wrong
export $(xargs < .vault/env/.env.local)
./gradlew testEnd2End
```

To validate how the builder container works (note that this is - at this point in time - not intended for developers to run on a regular basis, this container is - again, at this point in time - intended for the build pipeline)
```bash
$ cd prime-router # all runs from prime-router
# Stop everything docker (note: this will get rid of all docker containers and their configurations)
docker stop -t 1 <all_containers>  # or the equivalent necessary docker-compose downs if you have compose environments running
docker system prune --volumes -f  # Prune your system, including your volumes
docker-compose --file docker-compose.yml up  # Bring up the docker-compose development environment pre-requisites
#
# Actual validation; if you have a lot of time, reset your environment in between each using the steps above
# 
./build.sh -- gradle clean
./build.sh  # just a simple build (equivalent to -- gradle package)
./build.sh -- gradle clean package -x fatjar # Generate the large jar
./build.sh -- gradle generateDocs  # generate documentation
./build.sh -- gradle compile
./build.sh -- gradle migrate
./build.sh -- gradle test
```

### Testing
- [x] Tested locally?
- [ ] Ran `quickTest all`?
- [ ] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Security
- [x] Did you check for sensitive data, and remove any? 
- [x] Does logging contain sensitive data?  
- [ ] Are additional approvals needed for this change?
- [x] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

### Process
- [x] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [x] DevOps team has been notified if PR requires ops support?

## Security
*List potential security threats and mitigations in the PR*

## Fixes
*List GitHub issues this PR fixes*
- #1256 : Build the app in a Docker container

## To Be Done
- #1012: Integrate Docker-Compose plugin into Gradle

### CODE REVIEW

**Comment by carlosfelix2** [comment]
Can you provide a bit more info on this?  What is this for and for whom?  Use cases?

File: prime-router/build.sh
Comment (carlosfelix2) [inline prime-router/build.sh:29]: You need usage information here.

**Comment by td-usds** [comment]
> Can you provide a bit more info on this? What is this for and for whom? Use cases?

The purpose of this is to provide a standardized, reproducible environment in which we build. This is not about producing containers as output of the build and instead is about using containers during the build. This will ensure consistency in the produced artifacts (and make it easier to have an environment in which you can perform a build).

Secondly, as per #1012, it'd be great if all of this is neatly integrated with our current build system (Gradle) in such a way that any of the dependant (as opposed to dependent) container is spun up automatically when you initiate a build.

**Comment by carlosfelix2** [comment]
> > Can you provide a bit more info on this? What is this for and for whom? Use cases?
> 
> #1256: The purpose of this is to provide a standardized, reproducible environment in which we build. This is not about producing containers as output of the build and instead is about using containers during the build. This will ensure consistency in the produced artifacts (and make it easier to have an environment in which you can perform a build).
> 
> Secondly, as per #1012, it'd be great if all of this is neatly integrated with our current build system (Gradle) in such a way that any of the dependant (as opposed to dependent) container is spun up automatically when you initiate a build.

I guess, my confusion is that Gradle is supposed to 

> > Can you provide a bit more info on this? What is this for and for whom? Use cases?
> 
> #1256: The purpose of this is to provide a standardized, reproducible environment in which we build. This is not about producing containers as output of the build and instead is about using containers during the build. This will ensure consistency in the produced artifacts (and make it easier to have an environment in which you can perform a build).
> 
> Secondly, as per #1012, it'd be great if all of this is neatly integrated with our current build system (Gradle) in such a way that any of the dependant (as opposed to dependent) container is spun up automatically when you initiate a build.

I am not sure I am following this approach.  I wanted to run it locally, but I get an error on the script.  Can you add usage details on how to run this and what to expect?

File: .editorconfig
Comment (ronaldheft-usds) [inline .editorconfig:1]: Yes! We need this.

File: prime-router/docker-compose.build.yml
Comment (ronaldheft-usds) [inline prime-router/docker-compose.build.yml:19]: Any reason for specifying the DNS?

File: prime-router/docker-compose.build.yml
Comment (ronaldheft-usds) [inline prime-router/docker-compose.build.yml:16]: Nice! So any artifacts built in the container will be available outside the container.

**Review by ronaldheft-usds — COMMENTED** [review]
I'm liking this approach. It also provides the option that if someone does want to build outside of Docker, they can. We're not changing up the Gradle build, just providing a consistent environment to run it in.

I see you have a sample GitHub Action for building the application. One downside I see to the approach of building through Docker is we need to rebuild the builder image each time in CI. I'm wondering if we could improve that by publishing our building image to our repo as a package? As a public repo, we should have free, unlimited GitHub packages access.

File: prime-router/docker-compose.build.yml
Comment (td-usds) [reply prime-router/docker-compose.build.yml:16]: Exactly, and this is what makes it so that you can first build through the container, then just invoke `gradle[w]`, then through the container again, etc... make it as frictionles as possible :)

File: prime-router/docker-compose.build.yml
Comment (td-usds) [reply prime-router/docker-compose.build.yml:19]: Not really, this is residual to my own personal setup, this will be removed when the PR exists draft status :) I'll add a TODO so that any review can more easily pick it up
good catch :)

Comment (td-usds) [reply prime-router/docker-compose.build.yml:19]: I had a bit of a think about the pre-built images and I think it makes sense to do that. We can tag them accordingly so that we have the ability to quickly flip between other combinations of JDK/Gradle.

**Review by ronaldheft-usds — APPROVED** [review]
I completed a screenshare code review with @td-usds and have reviewed the code. I like this approach to running through the Docker, and we still have some flexibility with running Gradle outside Docker.

My only concern is the IP `172.17.0.1` and if it's cross platform across Docker environments (Windows, Mac, Linux). If it it's good on the three major platforms, and there aren't any typical reasons why a user would have a different default, then I'm good with it.

I'm approving, but this is something we should get a few more eyes on for an approval before we merge.

**Comment by MauriceReeves-usds** [comment]
@RickHawesUSDS - it would be good to get your feedback on this PR

File: frontend/eleventy.config.js
Comment (carlosfelix2) [inline frontend/eleventy.config.js:2]: Remove this commented line

File: prime-router/build.gradle.kts
Comment (carlosfelix2) [inline prime-router/build.gradle.kts:7]: This breaks running the code locally.  We should talk about how these changes are going to change development.

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

**Comment by carlosfelix2** [comment]
> I completed a screenshare code review with @td-usds and have reviewed the code. I like this approach to running through the Docker, and we still have some flexibility with running Gradle outside Docker.
> 
> My only concern is the IP `172.17.0.1` and if it's cross platform across Docker environments (Windows, Mac, Linux). If it it's good on the three major platforms, and there aren't any typical reasons why a user would have a different default, then I'm good with it.
> 
> I'm approving, but this is something we should get a few more eyes on for an approval before we merge.

I understand the goal, but I think this will break locally running of the code.  Let's talk about this implementation before we merge it in.

**Comment by td-usds** [comment]
@carlosfelix2 as long as you have docker installed, this should not make a difference and running locally should continue to work the way it did before.
`172.17.0.1` is the IP address associated with the Docker Bridge network Gateway. It is in the private IP range `172.16.0.0/12` and created by docker itself. Any container is automatically connected to the bridge network as well which means that any container can talk to it (unless explicitly NOT connected to it, but by default, you are connected to it). This IP address is also accessible from your local machine because that's what docker makes available locally.
``` ip add show docker0
n: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether <snip> brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 <snip>/64 scope link 
       valid_lft forever preferred_lft forever
```
```json
$ docker network inspect bridge
[
    {
        "Name": "bridge",
        "Id": "<snip>",
        "Created": "<snip>",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
        },
        // snip
    }
]
```

In this case, we're using it to reach the docker host machine (i.e. your 'real' machine) on which all other development environment containers are exposing their own ports as well.

Are there particular cases that you are concerned about that would not run locally anymore? Let me know and then I can put mitigations/fixes in for those.

File: frontend/eleventy.config.js
Comment (td-usds) [reply frontend/eleventy.config.js:2]: Good catch

File: prime-router/build.gradle.kts
Comment (td-usds) [reply prime-router/build.gradle.kts:7]: This should not break anything unless you do not have docker (which _is_ part of our toolchain anyway). Even if you run PostgreSQL not in docker (i.e. straight on your machine), this will still work (I just validated this with `nc`). To the best of my understanding, for all intents and purposes, `172.17.0.1` ends you up at the same place ends you at `127.0.0.1`.

**Comment by td-usds** [comment]
@carlosfelix2 may be on to something, I think we are hitting this issue: https://docs.docker.com/docker-for-windows/networking/

**Comment by td-usds** [comment]
I'm reverting this back to draft until we have a solution for non-Linux platforms as well

**Review by jimduff-usds — CHANGES_REQUESTED** [review]
Thomas, let's have a presentation of this at a dev sync session before you merge it in, with a focus on why we're making this change, and what developers need to do differently than before. Thanks.

I get this error on the 3rd step:

```
$ docker-compose --file docker-compose.yml up 
Creating network "prime-router_prime-router_build" with the default driver
Creating volume "prime-router_vault" with default driver
Creating prime-router_sftp_1    ... done
Creating prime-router_redox_1   ... done
Creating prime-router_vault_1     ... done
Creating prime-router_azurite_1 ... done
Creating prime-router_prime_dev_1 ... error

ERROR: for prime-router_prime_dev_1  Cannot start service prime_dev: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/prime-data-hub-router/start_func.sh": stat /prime-data-hub-router/start_func.sh: no such file or directory: unknown

ERROR: for prime_dev  Cannot start service prime_dev: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/prime-data-hub-router/start_func.sh": stat /prime-data-hub-router/start_func.sh: no such file or directory: unknown
ERROR: Encountered errors while bringing up the project.
```

`./build.sh `seems to require a password?   And if you ^C out of it, it seems to continue building anyway.

**Comment by td-usds** [comment]
> Thomas, let's have a presentation of this at a dev sync session before you merge it in, with a focus on why we're making this change, and what developers need to do differently than before. Thanks.

Yes, I've moved this PR back to draft because of the issue that @carlosfelix2 highlighted as well. Right now this works on Linux but not on Windows (and I haven't tried mac yet) due to some networking issues related to the docker bridge. So I have some work to do on this.

> I get this error on the 3rd step:
> 
> ```
> $ docker-compose --file docker-compose.yml up 
> Creating network "prime-router_prime-router_build" with the default driver
> Creating volume "prime-router_vault" with default driver
> Creating prime-router_sftp_1    ... done
> Creating prime-router_redox_1   ... done
> Creating prime-router_vault_1     ... done
> Creating prime-router_azurite_1 ... done
> Creating prime-router_prime_dev_1 ... error
> 
> ERROR: for prime-router_prime_dev_1  Cannot start service prime_dev: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/prime-data-hub-router/start_func.sh": stat /prime-data-hub-router/start_func.sh: no such file or directory: unknown
> 
> ERROR: for prime_dev  Cannot start service prime_dev: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/prime-data-hub-router/start_func.sh": stat /prime-data-hub-router/start_func.sh: no such file or directory: unknown
> ERROR: Encountered errors while bringing up the project.
> ```

I think this is related to the networking issue as well. Just for my information, can you share your configuration with me (i.e. OS, environment, ...) where you ran this? To the best of my current understanding, I think this is related to trying to bring something up that couldn't build.

> `./build.sh `seems to require a password? And if you ^C out of it, it seems to continue building anyway.

Correct, this is to be able to deal with having different UIDs within and outside of the docker container in which the build itself runs and make it as transparent as possible.
Basically, when you build in a clean repo and thus you don't have `build/`, `.gradle/`, nor `docs/`, then the build creates those for you as part of the process in the bind-mounted directory `prime-router`. But this creation happens _inside_ the container. Thus the owner of those directories is the user in which that process inside the container runs, which is `root` (within the container's namespace of course) and that user has UID 0. Thus anything created by that user has its owner set to 0 as well. These directories (and their content is created _inside_ the container but really live outside of the container and are bind-mounted in (or in reverse: it's a local directory mounted inside your container). The result of this is that you now have local directories inside your repo (created by the build) that has the user with UID 0 as their owner (which is `root` - and not/should not be the user under which you typically operate) and thus your 'outside-of-container-user' would not have permissions to those directories outside of the container. If you then run a gradle build outside of the container (which you still can), you'd get weird permission errors because the build cannot write or even access to those directories anymore.
To accommodate for this, there's a function that `chown`'s those directories back to the 'outside-of-container-you'. That is where the password prompt is coming from. `Ctrl+C` just prevents that `chown`ing from happening and building can happily continue within the container because there the process is running as 'inside-the-container-root' anyway. It's there as a safeguard and convenience for those who wish to build in a 'heterogeneous' way (i.e. both inside and outside of the container):
```bash
function ensure_build_dir() {
  mkdir -p "${HERE?}/build"
  chmod 777 "${HERE?}/build"
  echo "Making sure you own 'build/', 'docs/' and '.gradle/'... (may require elevation)"
  sudo chown -R "$(id --user --name):$(id --group --name)" \
    build \
    docs \
    .gradle
}
```

**Comment by ronaldheft-usds** [comment]
Just ran this and it looks the script itself succeeded, but still got some test failures:

```
Test passed: for ignore.CSV action receive:  Expecting 5 item lineage records and got 5
Test passed: for ignore.CSV action batch:  Expecting 5 item lineage records and got 5
Test passed: for ignore.CSV action send:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7 action receive:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7 action batch:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7 action send:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7_BATCH action receive:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7_BATCH action batch:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7_BATCH action send:  Expecting 5 item lineage records and got 5
Test passed: for ignore.REDOX action receive:  Expecting 5 item lineage records and got 5
Test passed: for ignore.REDOX action batch:  Expecting 5 item lineage records and got 5
Test passed: for ignore.REDOX action send:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7_NULL action receive:  Expecting 5 item lineage records and got 5
Test passed: for ignore.HL7_NULL action batch:  Expecting 5 item lineage records and got 5
*** TEST FAILED*** for ignore.OTC_PROCTORED_YYY action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.OTC_PROCTORED_NYY action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.OTC_PROCTORED_NUNKUNK action receive:  Expecting 5 item lineage records but got 0
*** Tests FAILED:  end2end ***
```

**Comment by td-usds** [comment]
> Just ran this and it looks the script itself succeeded, but still got some test failures:
> 
> ```
> Test passed: for ignore.CSV action receive:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.CSV action batch:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.CSV action send:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7 action receive:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7 action batch:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7 action send:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7_BATCH action receive:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7_BATCH action batch:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7_BATCH action send:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.REDOX action receive:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.REDOX action batch:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.REDOX action send:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7_NULL action receive:  Expecting 5 item lineage records and got 5
> Test passed: for ignore.HL7_NULL action batch:  Expecting 5 item lineage records and got 5
> *** TEST FAILED*** for ignore.OTC_PROCTORED_YYY action receive:  Expecting 5 item lineage records but got 0
> *** TEST FAILED*** for ignore.OTC_PROCTORED_NYY action receive:  Expecting 5 item lineage records but got 0
> *** TEST FAILED*** for ignore.OTC_PROCTORED_NUNKUNK action receive:  Expecting 5 item lineage records but got 0
> *** Tests FAILED:  end2end ***
> ```

That is due to #1583

File: .github/workflows/build_hub.yml
Comment (ronaldheft-usds) [inline .github/workflows/build_hub.yml:46]: 👀

**Review by ronaldheft-usds — APPROVED** [review]
This looks great to me, and I like all the extra documentation for installing various base tools like Azure and gradle.

**Comment by jimduff-usds** [comment]
@td-usds, this needed review and tested by someone doing active development and QA.  Our ReportStream tests are not working now.

Our ./quick-test.sh script now asks for a password (???)
Also my local ./prime test smoktests are failing:

*** TEST FAILED*** for ignore.CSV action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.CSV action batch:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.CSV action send:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7 action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7 action batch:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7 action send:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7_BATCH action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7_BATCH action batch:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7_BATCH action send:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.REDOX action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.REDOX action batch:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.REDOX action send:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7_NULL action receive:  Expecting 5 item lineage records but got 0
*** TEST FAILED*** for ignore.HL7_NULL action batch:  Expecting 5 item lineage records but got 0

I think you need to demo at a DevSync before merging.

**Pull request merged**
